src/main/scala/beam/agentsim/infrastructure/HierarchicalParkingManager.scala

```diff
@@ -0,0 +1,253 @@
+import beam.agentsim.infrastructure.HierarchicalParkingManager.{ParkingCluster, Worker}
+import beam.agentsim.infrastructure.parking.ParkingZone
+import beam.agentsim.infrastructure.parking.ParkingZoneSearch.ZoneSearchTree
+import beam.agentsim.infrastructure.taz.{TAZ, TAZTreeMap}
+import beam.sim.common.GeoUtils
+import beam.sim.common.GeoUtils.toJtsCoordinate
+import beam.sim.config.BeamConfig
+import com.typesafe.scalalogging.LazyLogging
+import com.vividsolutions.jts.algorithm.ConvexHull
+import com.vividsolutions.jts.geom.{Coordinate, Envelope, Geometry, GeometryFactory}
+import de.lmu.ifi.dbs.elki.algorithm.clustering.kmeans.KMeansElkan
+import de.lmu.ifi.dbs.elki.algorithm.clustering.kmeans.initialization.RandomUniformGeneratedInitialMeans
+import de.lmu.ifi.dbs.elki.data.`type`.TypeUtil
+import de.lmu.ifi.dbs.elki.data.{DoubleVector, NumberVector}
+import de.lmu.ifi.dbs.elki.database.ids.DBIDIter
+import de.lmu.ifi.dbs.elki.database.relation.Relation
+import de.lmu.ifi.dbs.elki.database.{Database, StaticArrayDatabase}
+import de.lmu.ifi.dbs.elki.datasource.ArrayAdapterDatabaseConnection
+import de.lmu.ifi.dbs.elki.distance.distancefunction.minkowski.SquaredEuclideanDistanceFunction
+import de.lmu.ifi.dbs.elki.utilities.random.RandomFactory
+import org.matsim.api.core.v01.{Coord, Id}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.ExecutionContext
+import scala.util.Random
+
+/**
+  *
+  * @author Dmitry Openkov
+  */
+class HierarchicalParkingManager(
+  beamConfig: BeamConfig,
+  tazTreeMap: TAZTreeMap,
+  clusters: Vector[ParkingCluster],
+  zones: Array[ParkingZone],
+  searchTree: ZoneSearchTree[TAZ],
+  geo: GeoUtils,
+  rand: Random,
+  boundingBox: Envelope
+) extends beam.utils.CriticalActor
+    with ActorLogging {
+  private implicit val timeout: Timeout = Timeout(100, TimeUnit.SECONDS)
+  private implicit val executionContext: ExecutionContext = context.dispatcher
+
+  private val workers: Vector[Worker] =
+    clusters.zipWithIndex.map {
+      case (cluster, i) =>
+        val emergencyId = Id.create(s"emergency-$i", classOf[TAZ])
+        val actor = context.actorOf(
+          ZonalParkingManager
+            .props(
+              beamConfig,
+              TAZTreeMap.fromSeq(cluster.tazes),
+              zones,
+              searchTree,
+              emergencyId,
+              geo,
+              rand,
+              boundingBox
+            ).withDispatcher("hierarchical-parking-manager-dispatcher"),
+          s"zonal-parking-manager-$i"
+        )
+        Worker(actor, cluster, emergencyId)
+    }
+
+  private val tazToWorker: Map[Id[TAZ], Worker] = mapTazToWorker(workers)
+
+  override def receive = {
+    case inquiry: ParkingInquiry =>
+      val foundCluster = workers.find { w =>
+        val point = HierarchicalParkingManager.geometryFactory.createPoint(inquiry.destinationUtm)
+        w.cluster.convexHull.contains(point)
```

Comment:
What is the number of workers/tazs?
QuadTree querying would have better complexity comparing to iterative search by contains.
Also i would advice to wrap `convexHull` into `PreparedGeometry` class which is enhanced to be fast for operations like **contains** and **intersects**.

Reply:
The default number of workers is 8. Each worker can have 1 to 500 (or more) to tazs in its cluster.

Reply:
I changed the code so that it uses PreparedGeometry. About QuadTree querying: we need probably create a benchmark to see what has better performance.